src/main/java/io/r2dbc/h2/codecs/ArrayCodec.java

```diff
@@ -47,6 +47,6 @@ boolean doCanDecode(int dataType) {
     Value doEncode(Object[] value) {
         return ValueArray.get(Arrays.stream(Assert.requireNonNull(value, "value must not be null"))
             .map(codecs::encode)
-            .toArray(Value[]::new));
+            .toArray(Value[]::new), null);
```

Comment:
H2 2.0 has only standard-compliant typed arrays.

`CastDataProvider` shouldn't be `null` here. If objects of different types are passed within the same array, they  need to be converted to the same higher type and some conversions need to know properties of the session.

Reply:
I've opened https://github.com/r2dbc/r2dbc-h2/issues/208 to track this.